GPT2 sample failing with the new graph engine

### System Information

OpenCV version: 5.x
Operating System / Platform: MacOS



### Detailed description

Running gpt2 inference sample fail if the sequence length is different comapred to export  sequence length. This behaviour is present using new graph engine (it should suppport dynamic shapes). 

Error is 
```bash
logits = net.forward()
cv2.error: OpenCV(5.0.0-pre) /Users/abd/work/opencv/modules/dnn/src/layers/nary_eltwise_layers.cpp:308: error: (-215:Assertion failed) shape[i] == 1 || outShape[i] == 1 in function 'findCommonShape'
```

### Steps to reproduce

Run[ gpt2 inference sample](https://github.com/opencv/opencv/blob/5.x/samples/dnn/gpt2_inference.py) as described in the sample. Include [this patch](https://github.com/opencv/opencv/pull/26514#issuecomment-2497024212) for the running the sample with new engine

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/samples/dnn/gpt2_inference.py
+++ b/samples/dnn/gpt2_inference.py
@@ -10,7 +10,7 @@ To export GPT-2 model to ONNX, you can use the following procedure:
 
 1. Clone fork of Andrej Karpathy's GPT-2 repository:
 
-    git clone https://github.com/Abdurrahheem/build-nanogpt/tree/ash/export-gpt2-onnx
+    git clone https://github.com/Abdurrahheem/build-nanogpt/tree/ash/export-gpt2-onnx-dynamic
 
 2. Install the required dependencies:
 
@@ -18,22 +18,20 @@ To export GPT-2 model to ONNX, you can use the following procedure:
 
 3  Export the model to ONNX:
 
-    python export2onnx.py --promt=<Any-promt-you-want> --batch_size=<batch-size>
+    python export2onnx.py --promt=<Any-promt-you-want>
 
 
 Run the script:
 1. Install the required dependencies:
 
-    pip install tiktoken==0.7.0
+    pip install tiktoken==0.7.0, numpy, tqdm
 
 2. Run the script:
-
-    python gpt2_inference.py --model=<path-to-onnx-model> --max_seq_len=<max-output-lenght> --batch_size=<use-one-used-while-exportinh> --prompt=<use-promt-of-the-same-length-used-while-exporting>
+    python gpt2_inference.py --model=<path-to-onnx-model>  --prompt=<use-promt-of-the-same-length-used-while-exporting>
 '''
 
 
 
-from copy import deepcopy
 import numpy as np
 import tiktoken
 import argparse
@@ -44,9 +42,9 @@ def parse_args():
     parser = argparse.ArgumentParser(description='Use this script to run GPT-2 inference in OpenCV',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
     parser.add_argument('--model', type=str, required=True, help='Path to GPT-2 model ONNX model file.')
-    parser.add_argument("--max_seq_len", type=int, default=30, help="Number of tokens to continue.")
-    parser.add_argument("--batch_size", type=int, default=5, help="Number of batches.")
     parser.add_argument("--prompt", type=str, default="Hello, I'm a language model,", help="Prompt to start with.")
+    parser.add_argument("--max_seq_len", type=int, default=40, help="Number of tokens to continue.")
+    parser.add_argument("--batch_size", type=int, default=1, help="Number of batches.")
     parser.add_argument("--seed", type=int, default=0, help="Random seed")
     return parser.parse_args()
 
@@ -55,7 +53,7 @@ def stable_softmax(logits):
     return exp_logits / np.sum(exp_logits, axis=-1, keepdims=True)
 
 
-def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
+def gpt2_inference(net, tokens, max_length, num_return_sequences=1):
 
     print("Inferencing GPT-2 model...")
     x = np.array(tokens)

--- a/samples/dnn/gpt2_inference.py
+++ b/samples/dnn/gpt2_inference.py
@@ -10,7 +10,7 @@ To export GPT-2 model to ONNX, you can use the following procedure:
 
 1. Clone fork of Andrej Karpathy's GPT-2 repository:
 
-    git clone https://github.com/Abdurrahheem/build-nanogpt/tree/ash/export-gpt2-onnx-dynamic
+    git clone -b ash/export-gpt2-onnx-dynamic https://github.com/Abdurrahheem/build-nanogpt.git
 
 2. Install the required dependencies:
 
@@ -24,7 +24,7 @@ To export GPT-2 model to ONNX, you can use the following procedure:
 Run the script:
 1. Install the required dependencies:
 
-    pip install tiktoken==0.7.0, numpy, tqdm
+    pip install tiktoken==0.7.0 numpy tqdm
 
 2. Run the script:
     python gpt2_inference.py --model=<path-to-onnx-model>  --prompt=<use-promt-of-the-same-length-used-while-exporting>

--- a/samples/dnn/gpt2_inference.py
+++ b/samples/dnn/gpt2_inference.py
@@ -38,6 +38,7 @@ import numpy as np
 import tiktoken
 import argparse
 import cv2 as cv
+from tqdm import tqdm
 
 def parse_args():
     parser = argparse.ArgumentParser(description='Use this script to run GPT-2 inference in OpenCV',
@@ -58,13 +59,17 @@ def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
 
     print("Inferencing GPT-2 model...")
     x = np.array(tokens)
-    x = np.tile(x, (num_return_sequences, 1))
+    x = np.tile(x, (num_return_sequences, 1)).astype(np.int32)
+    pos = np.arange(0, len(x), dtype=np.int32)
 
-    output_buffer = deepcopy(x)
     counter = x.shape[1]
+    pbar = tqdm(total=max_length - counter, desc="Generating tokens")
     while counter < max_length:
 
-        net.setInput(x)
+        net.setInputsNames(['input_ids', 'position_ids'])
+        net.setInput(x, 'input_ids')
+        net.setInput(pos, 'position_ids')
+
         logits = net.forward()
 
         # logits is assumed to be (B, seq_length, vocab_size) and needs to be the last token's logits
@@ -86,12 +91,14 @@ def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
 
         # Append to the sequence
         x = np.concatenate((x, sampled_indices), axis=1)
-        x = x[:, 1:] ## issue due to fixes size window in opencv
+        pos = np.arange(0, x.shape[1], dtype=np.int32) # shape (T)
 
-        output_buffer = np.concatenate((output_buffer, sampled_indices), axis=1)
         counter += 1
+        pbar.update(1)
+
+    pbar.close()
     print("Inference done!")
-    return output_buffer
+    return x
 
 if __name__ == '__main__':
 
@@ -102,24 +109,13 @@ if __name__ == '__main__':
     prompt = args.prompt
 
     net = cv.dnn.readNet(args.model)
-    input_token_size = net.getLayerShapes([], 0, 0)[0][0][1]
 
     enc = tiktoken.get_encoding('gpt2')
     tokens = enc.encode(prompt)
 
-    # Check if the prompt is of the same length as the input tokens
-    # if not, pad the tokens else truncate the tokens
-    if len(tokens) > input_token_size:
-        tokens = tokens[:input_token_size]
-    elif len(tokens) < input_token_size:
-        tokens2pad = input_token_size - len(tokens)
-        # append <space> token to the prompt
-        tokens += [220] * tokens2pad
-
-
-    output_buffer = gpt2_inference(net, tokens, max_length, num_return_sequences)
+    output = gpt2_inference(net, tokens, max_length, num_return_sequences)
 
     for i in range(num_return_sequences):
-        tokens = output_buffer[i, :max_length].tolist()
+        tokens = output[i].tolist()
         decoded = enc.decode(tokens)
         print(">>>>", decoded)

